libcylinder/src/jwt/mod.rs

```diff
@@ -14,3 +14,504 @@
+    ///  ```
+    ///
+    /// # Errors
+    ///
+    /// A `JsonWebTokenBuildError` may be returned if the token can not be properly built or
+    /// signed.
+    pub fn build(self, signer: &dyn Signer) -> Result<String, JsonWebTokenBuildError> {
+        let mut jwt_header = json::JsonValue::new_object();
+
+        for (k, v) in self.header {
+            jwt_header[k] = v.into();
+        }
+        jwt_header["alg"] = "secp256k1".into();
+        jwt_header["typ"] = "cylinder+jwt".into();
+
+        // Header bytes are UTF-8 bytes of the JSON string representation of the header
+        let header_bytes = json::stringify(jwt_header).into_bytes();
+
+        let public_key = signer.public_key().map_err(|e| {
+            JsonWebTokenBuildError::new(
+                "Unable to get the public key from the provided signer".into(),
+                e,
+            )
+        })?;
+
+        let mut claims = json::JsonValue::new_object();
+        for (k, v) in self.claims {
+            claims[k] = v.into()
+        }
+
+        claims["iss"] = public_key.as_hex().into();
+
+        // Claims bytes are the UTF-8 bytes of the JSON string representation of the header
+        let claims_bytes = json::stringify(claims).into_bytes();
+
+        let mut token = String::new();
+
+        token.push_str(&base64::encode(header_bytes));
+        token.push('.');
+        token.push_str(&base64::encode(claims_bytes));
+
+        let signature = signer
+            .sign(token.as_bytes())
+            .map_err(|e| JsonWebTokenBuildError::new("Unable to sign the token".into(), e))?;
+
+        token.push('.');
+        token.push_str(&base64::encode(signature.as_slice()));
+
+        Ok(token)
+    }
+}
+
+/// Parses JsonWebToken struct from an encoded token.  The token's signature is verified before the
+/// parsed value is returned.
+pub struct JsonWebTokenParser<'a> {
+    verifier: &'a dyn Verifier,
+}
+
+impl<'a> JsonWebTokenParser<'a> {
+    /// Construct a new parser instance around the given verifier.
+    pub fn new(verifier: &'a dyn Verifier) -> Self {
+        Self { verifier }
+    }
+
+    /// Parse the token string provided and verify the included signature
+    /// with the parser's Verifier instance.
+    ///
+    /// # Errors
+    ///
+    /// A `JsonWebTokenParseError` may be returned if the token is no properly formed or the
+    /// signature is not valid.
+    pub fn parse(&self, jwt_string: &str) -> Result<JsonWebToken, JsonWebTokenParseError> {
+        let mut encoded_token_parts = jwt_string.split('.');
+        let (encoded_header, encoded_claims, encoded_signature) = {
+            match (
+                encoded_token_parts.next(),
+                encoded_token_parts.next(),
+                encoded_token_parts.next(),
+            ) {
+                (Some(encoded_header), Some(encoded_claims), Some(encoded_signature)) => {
+                    (encoded_header, encoded_claims, encoded_signature)
+                }
```

Comment:
This should include a `(None, _, _) => unreachable!(),` branch, which is separate from the Ok and Err conditions (this is based on recent conversations with @vaporos about being as clear as possible with these matches).

Reply:
Fixed